packages/stream-metadata/src/aws.ts

```diff
@@ -1,37 +1,106 @@
-import { CloudFront } from '@aws-sdk/client-cloudfront'
+import { CloudFront, GetInvalidationCommandOutput } from '@aws-sdk/client-cloudfront'
 import { FastifyBaseLogger } from 'fastify'
 
-import { config } from './environment'
+import { config as envConfig } from './environment'
 
-export const createCloudfrontInvalidation = async (params: {
-	path: string
-	logger: FastifyBaseLogger
-}) => {
-	if (!config.aws?.CLOUDFRONT_DISTRIBUTION_ID) {
-		params.logger.warn(
+export class CloudfrontManager {
+	private readonly cloudFront: CloudFront
+
+	constructor(
+		private readonly logger: FastifyBaseLogger,
+		private readonly config: {
+			distributionId: string
+			invalidationConfirmationMaxAttempts: number
+			invalidationConfirmationWait: number
+		},
+	) {
+		this.cloudFront = new CloudFront({
+			serviceId: 'stream-metadata',
+			logger: this.logger,
+		})
+	}
+
+	private async invalidate(params: { path: string; waitUntilFinished?: boolean }) {
+		const invalidationCommand = await this.cloudFront.createInvalidation({
+			DistributionId: this.config.distributionId,
+			InvalidationBatch: {
+				CallerReference: `${new Date().toISOString()}-${params.path.substring(0, 5)}`,
+				Paths: {
+					Quantity: 1,
+					Items: [params.path],
+				},
+			},
+		})
+
+		this.logger.info({ path: params.path }, 'CloudFront cache invalidation created')
+
+		if (params.waitUntilFinished) {
+			await this.waitForInvalidation(invalidationCommand, params.path)
+		}
+	}
+
+	private async waitForInvalidation(
+		invalidationCommand: GetInvalidationCommandOutput,
+		path: string,
+	) {
+		let attempts = 0
+		let currentInvalidationCommand = invalidationCommand
+		while (currentInvalidationCommand.Invalidation?.Status !== 'Completed') {
+			attempts += 1
+			if (attempts >= this.config.invalidationConfirmationMaxAttempts) {
+				this.logger.error(
+					{
+						invalidation: currentInvalidationCommand,
+						path,
+					},
+					'CloudFront cache invalidation did not complete in time',
+				)
+				throw new Error('CloudFront cache invalidation did not complete in time')
+			}
+			this.logger.info(
+				{
+					invalidation: currentInvalidationCommand,
+					path,
+				},
+				'Waiting for CloudFront cache invalidation to complete...',
+			)
+			if (!currentInvalidationCommand.Invalidation) {
+				throw new Error('Invalidation not found')
+			}
+			await new Promise((resolve) =>
+				setTimeout(resolve, this.config.invalidationConfirmationWait),
```

Comment:
would be nice to do this without the sleep, but even [aws sdk does that](https://github.com/aws/aws-sdk-js-v3/blob/e6cba699ea7ca7d9c928e1eae17fc4089e985676/clients/client-s3/src/waiters/waitForObjectExists.ts#L30) (check the [waiter impl](https://github.com/smithy-lang/smithy-typescript/blob/main/packages/util-waiter/src/poller.ts#L25))